test/integration/api/personal/access_tokens_test.rb

```diff
@@ -0,0 +1,199 @@
+      another_admin = FactoryBot.create(:simple_admin, account: @provider)
+      [@admin, another_admin].each do |owner|
+        %w[searchable another_name].each { |name| FactoryBot.create_list(:access_token, 2, owner: owner, name: name) }
+      end
+      perform_request(different_params: {name: 'arch'})
+      assert_it_worked do |access_tokens_json_response|
+        response_ids = (access_tokens_json_response.map { |access_token| access_token.dig('access_token', 'id') })
+        assert_same_elements(@admin.access_tokens.where('name LIKE \'%arch%\'').pluck(:id), response_ids)
+      end
+    end
+
+    def perform_request(authentication: {access_token: @admin_access_token.value}, different_params: {})
+      get admin_api_personal_access_tokens_path(authentication.merge(different_params))
+    end
+
+    def assert_it_worked
+      assert_response :ok
+      assert_instance_of Array, (access_tokens_json_response = JSON.parse(response.body)['access_tokens'])
+      yield(access_tokens_json_response) if block_given?
+    end
+  end
+
+  test 'authenticate with provider_key is forbidden' do
+    FactoryBot.create(:cinstance, service: master_account.default_service, user_account: @provider)
+
+    assert_no_difference(AccessToken.method(:count)) do
+      perform_request(authentication: {provider_key: @provider.provider_key})
+      assert_response :forbidden
+    end
+  end
+
+  test 'authentication forbidden for member user with the wrong permissions' do
+    unauthorized_member = FactoryBot.create(:member, account: @provider, admin_sections: [])
+    access_token = FactoryBot.create(:access_token, owner: unauthorized_member, scopes: %w[account_management])
+
+    assert_no_difference(AccessToken.method(:count)) do
+      perform_request(authentication: {access_token: access_token.value})
+      assert_response :forbidden
+    end
+  end
+
+  test 'authentication forbidden for member user with the right permissions but wrong access token' do
+    authorized_member = FactoryBot.create(:member, account: @provider, admin_sections: [:partners])
+    wrong_token_scope = FactoryBot.create(:access_token, owner: authorized_member, scopes: %w[finance])
+
+    assert_no_difference(AccessToken.method(:count)) do
+      perform_request(authentication: {access_token: wrong_token_scope.value})
+      assert_response :forbidden
+    end
+  end
+
+  test 'authentication allowed for account_management access token belonging to a member user with the right permissions' do
+    authorized_member = FactoryBot.create(:member, account: @provider, admin_sections: [:partners])
+    access_token = FactoryBot.create(:access_token, owner: authorized_member, scopes: %w[account_management])
+    @access_token_object = FactoryBot.create(:access_token, owner: authorized_member)
+
+    perform_request(authentication: {access_token: access_token.value})
+    assert_it_worked
+  end
+
+  def perform_request(*)
+    skip 'Need to be implemented in its subclasses'
```

Comment:
`skip` or `fail` ?

Reply:
I had no idea of the existence of this `flunk` until you mentioned. Here it says that it just fails:
https://apidock.com/ruby/MiniTest/Assertions/flunk
And that's not what I want, I want it to execute if it is implemented, which it is in all the subclasses 😄 Right now it only skips in the main class which is executed on its own as well unfortunately.

Reply:
What you can do then is to use modules instead of inheritance.
The only thing is you would not be able to use `test` DSL :/